config-controller/pkg/client/client.go

```diff
@@ -0,0 +1,303 @@
+	GetPolicy(ctx context.Context, id string) (*storage.Policy, error)
+	PostPolicy(context.Context, *storage.Policy) (*storage.Policy, error)
+	PutPolicy(context.Context, *storage.Policy) error
+	TokenExchange(ctx context.Context) error
+}
+
+type grpcClient struct {
+	svc         v1.PolicyServiceClient
+	perRPCCreds *perRPCCreds
+}
+
+func newGrpcClient(ctx context.Context) (PolicyClient, error) {
+	clientconn.SetUserAgent(clientconn.ConfigController)
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithNoProxy(),
+	}
+
+	perRPCCreds := &perRPCCreds{}
+	opts := clientconn.Options{
+		InsecureNoTLS:                  false,
+		InsecureAllowCredsViaPlaintext: false,
+		DialOptions:                    dialOpts,
+		PerRPCCreds:                    perRPCCreds,
+	}
+
+	callOpts := []grpc.CallOption{grpc.MaxCallRecvMsgSize(12 * size.MB)}
+
+	conn, err := clientconn.GRPCConnection(ctx, mtls.CentralSubject, centralHostPort, opts, grpc.WithDefaultCallOptions(callOpts...))
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create gRPC connection")
+	}
+
+	svc := v1.NewPolicyServiceClient(conn)
+	perRPCCreds.svc = v1.NewAuthServiceClient(conn)
+
+	return &grpcClient{
+		perRPCCreds: perRPCCreds,
+		svc:         svc,
+	}, nil
+}
+
+func (gc *grpcClient) ListPolicies(ctx context.Context) ([]*storage.ListPolicy, error) {
+	allPolicies, err := gc.svc.ListPolicies(ctx, &v1.RawQuery{})
+	if err != nil {
+		return []*storage.ListPolicy{}, errors.Wrap(err, "Failed to list policies from grpc client")
+	}
+
+	return allPolicies.Policies, err
+}
+
+func (gc *grpcClient) GetPolicy(ctx context.Context, id string) (*storage.Policy, error) {
+	policy, err := gc.svc.GetPolicy(ctx, &v1.ResourceByID{Id: id})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to fetch policy %s", id)
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) PostPolicy(ctx context.Context, policy *storage.Policy) (*storage.Policy, error) {
+	req := &v1.PostPolicyRequest{
+		Policy:                 policy,
+		EnableStrictValidation: true,
+	}
+
+	policy, err := gc.svc.PostPolicy(ctx, req)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to POST policy")
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) PutPolicy(ctx context.Context, policy *storage.Policy) error {
+	_, err := gc.svc.PutPolicy(ctx, policy)
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to PUT policy")
+	}
+
+	return nil
+}
+
+func (gc *grpcClient) TokenExchange(ctx context.Context) error {
+	if time.Since(gc.perRPCCreds.lastUpdated).Seconds() > 60.0 {
+		return gc.perRPCCreds.refreshToken(ctx)
+	}
+	return nil
+}
+
+type client struct {
+	svc         PolicyClient
+	cache       map[string]*storage.Policy
+	lastUpdated time.Time
+}
+
+type clientOptions interface {
+	Apply(CachedPolicyClient)
+}
+
+func New(ctx context.Context, opts ...clientOptions) (CachedPolicyClient, error) {
+	c := client{}
+
+	for _, o := range opts {
+		o.Apply(&c)
+	}
+
+	if c.svc == nil {
+		var gc PolicyClient
```

Comment:
no need to use extra variable gc. Use c.svc instead directly.

Reply:
aa3bd6a5a7ad0d3e2e7bfa34353161df3c98f602